TypeCobol/Compiler/Diagnostics/CrossChecker.cs

```diff
@@ -239,6 +239,83 @@ public override bool Visit(Move move)
             return true;
         }
 
+        public override bool Visit(WhenSearch whenSearch)
+        {
+            var search = whenSearch.Parent;
+            System.Diagnostics.Debug.Assert(search is Search);
+            if (search.CodeElement is SearchBinaryStatement)
+            {
+                var whenSearchVisitor = new WhenSearchVisitor(whenSearch);
+                whenSearch.CodeElement.Condition.AcceptASTVisitor(whenSearchVisitor);
+                if (whenSearchVisitor.IsInError)
+                {
+                    DiagnosticUtils.AddError(whenSearch, "First index declared for the table when subscripting and one of declared keys must be used.");
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Visit a WhenSearch and set a flag to true if either:
+        /// WHEN does not use one of the declared keys (ascending or descending)
+        /// WHEN does not use the first index declared for the table when subscripting
+        /// </summary>
+        private class WhenSearchVisitor : AbstractAstVisitor
+        {
+            private readonly WhenSearch _whenSearch;
+            public bool IsInError { get; private set; } = false;
+
+
+            public WhenSearchVisitor(WhenSearch whenSearch)
+            {
+                _whenSearch = whenSearch;
+            }
+
+
+            public override bool Visit(NumericVariable numericVariable)
+            {
+                var variableStorageArea = (DataOrConditionStorageArea)numericVariable.StorageArea;
+                // TODO : Handle the case of sub-tables
+                if (variableStorageArea?.Subscripts.Count == 1)
+                {
+                    ////////// WHEN must use one of the declared keys (ascending or descending) //////////
+                    var dataDefinitionKey = _whenSearch.GetDataDefinitionFromStorageAreaDictionary(variableStorageArea);
+                    if (dataDefinitionKey?.Parent.CodeElement is DataDescriptionEntry tableDescriptionEntry)
+                    {
+                        if (tableDescriptionEntry.TableSortingKeys != null)
+                        {
+                            var tableSortingKeyNames = tableDescriptionEntry.TableSortingKeys
```

Comment:
I think this can be simplified as follow:
```csharp
var existsName = tableDescriptionEntry.TableSortingKeys.SingleOrDefault(key => key.SortDirection?.Value != SortDirection.None && key.SortKey.Name.Equals(dataDefinitionKey.Name));
   if (existsName == null)
   {
   // error
   IsInError = true;
   }

```

Reply:
Your suggestion has been used with the `Any()` method from Linq instead of `SingleOrDefault()` as the result is not used.